liana-gui/src/app/settings/mod.rs

```diff
@@ -111,6 +114,184 @@ where
+        let params = Params::new(19456, 2, 1, None).expect("Valid Argon2 params");
+        let argon2 = Argon2::new(argon2::Algorithm::Argon2id, argon2::Version::V0x13, params);
+
+        // Hash the PIN
+        let password_hash = argon2
+            .hash_password(pin.as_bytes(), &salt)
+            .expect("Failed to hash PIN");
+
+        // Return the PHC string format
+        password_hash.to_string()
+    }
+
+    /// Verify PIN against Argon2id hash
+    fn verify_argon2_pin(pin: &str, hash: &str) -> bool {
+        use argon2::{
+            password_hash::{PasswordHash, PasswordVerifier},
+            Argon2,
+        };
+
+        // Parse the PHC string
+        let parsed_hash = match PasswordHash::new(hash) {
+            Ok(h) => h,
+            Err(_) => return false,
+        };
+
+        // Verify the PIN
+        Argon2::default()
+            .verify_password(pin.as_bytes(), &parsed_hash)
+            .is_ok()
+    }
+}
+
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct Cubes {
+    pub cubes: Vec<CubeSettings>,
+}
+
+impl Cubes {
+    pub fn from_file(network_dir: &NetworkDirectory) -> Result<Cubes, SettingsError> {
+        let mut path = network_dir.path().to_path_buf();
+        path.push(CUBES_FILE_NAME);
+
+        std::fs::read(path)
+            .map_err(|e| match e.kind() {
+                std::io::ErrorKind::NotFound => SettingsError::NotFound,
+                _ => SettingsError::ReadingFile(format!("Reading cubes file: {}", e)),
+            })
+            .and_then(|file_content| {
+                serde_json::from_slice::<Cubes>(&file_content)
+                    .map_err(|e| SettingsError::ReadingFile(format!("Parsing cubes file: {}", e)))
+            })
+    }
+}
+
+pub async fn update_cubes_file<F>(
+    network_dir: &NetworkDirectory,
+    updater: F,
+) -> Result<(), SettingsError>
+where
+    F: FnOnce(Cubes) -> Cubes,
+{
+    let path = network_dir.path().join(CUBES_FILE_NAME);
+    let file_exists = tokio::fs::try_exists(&path).await.unwrap_or(false);
+
+    let mut file = OpenOptions::new()
+        .read(true)
+        .write(true)
+        .create(true)
+        .truncate(false)
+        .open(&path)
+        .await
+        .map_err(|e| SettingsError::ReadingFile(format!("Opening file: {}", e)))?
+        .lock_write()
+        .await
+        .map_err(|e| SettingsError::ReadingFile(format!("Locking file: {:?}", e)))?;
+
+    let cubes = if file_exists {
+        let mut file_content = Vec::new();
+        file.read_to_end(&mut file_content)
+            .await
+            .map_err(|e| SettingsError::ReadingFile(format!("Reading file content: {}", e)))?;
+
+        serde_json::from_slice::<Cubes>(&file_content)
+            .map_err(|e| SettingsError::ReadingFile(e.to_string()))?
+    } else {
+        Cubes::default()
+    };
+
+    let cubes = updater(cubes);
+
+    if cubes.cubes.is_empty() {
```

Comment:
Here, instead of checking if cubes is empty (those pesky zero values I was talking about) you could have `updater` return `Option<Vec<CubeSettings>>`. If it returns None, the file is deleted. `Some([]) != None`